celery/canvas.py

```diff
@@ -550,86 +550,97 @@ def set(self, immutable=None, **options):
         # This will lose duplicated headers from the headers argument, but that is the best effort solution
         # to avoid implicitly casting the duplicated stamp into a list of both stamps from headers and
         # visitor_headers of the same key.
         # Example:
         #   headers = {"foo": "bar1"}
         #   visitor_headers = {"foo": "bar2"}
-        #   _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=True)
+        #   _merge_dictionaries(headers, visitor_headers, append_stamps=True)
         #   headers["foo"] == ["bar1", "bar2"] -> The stamp is now a list
-        #   _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=False)
+        #   _merge_dictionaries(headers, visitor_headers, append_stamps=False)
         #   headers["foo"] == "bar2" -> "bar1" is lost, but the stamp is according to the visitor
-        aggregate_duplicates = False
 
         headers = headers.copy()
+
+        if "stamped_headers" not in headers:
+            headers["stamped_headers"] = list(headers.keys())
+
         # Merge headers with visitor headers
         if visitor_headers is not None:
             visitor_headers = visitor_headers or {}
             if "stamped_headers" not in visitor_headers:
                 visitor_headers["stamped_headers"] = list(visitor_headers.keys())
-            # Prioritize visitor_headers over headers
-            _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=aggregate_duplicates)
+
+            # Sync from visitor
+            _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=append_stamps)
             headers["stamped_headers"] = list(set(headers["stamped_headers"]))
+
         # Merge headers with self.options
-        else:
-            headers["stamped_headers"] = [
-                header for header in headers.keys()
-                if header not in self.options and header != "stamped_headers"
-            ]
-
-            # Prioritize self.options over headers
-            _merge_dictionaries(headers, self.options, aggregate_duplicates=aggregate_duplicates)
-
-        # Sync missing stamps from self.options (relevant for stamping during task replacement)
-        stamped_headers = set(headers.get("stamped_headers", []))
-        stamped_headers.update(self.options.get("stamped_headers", []))
-        headers["stamped_headers"] = list(stamped_headers)
-        for previous_header in stamped_headers:
-            if previous_header not in headers and previous_header in self.options:
-                headers[previous_header] = self.options[previous_header]
+        if self_headers:
+            stamped_headers = set(headers.get("stamped_headers", []))
+            stamped_headers.update(self.options.get("stamped_headers", []))
+            headers["stamped_headers"] = list(stamped_headers)
+            reducted_options = {k: v for k, v in self.options.items() if k in headers["stamped_headers"]}
```

Comment:
Did you mean `redacted_options`?

Reply:
Also, why are they redacted? What do you mean by that?

Reply:
> Did you mean `redacted_options`?

Late night typos :)
Will fix.

Reply:
> Also, why are they redacted? What do you mean by that?

`{k: v for k, v in self.options.items() if k in headers["stamped_headers"]}`
Meaning `self.options` fields which are stamps.

Reply:
Added a comment to clarify + fixed typo - fix pushed